### ISSUE TITLE
Support ByteBuffer as a backing storage on JVM

### ISSUE THREAD

**Issue by fzhinkin** [Kotlin/kotlinx-io#239]
`java.nio.ByteBuffer` is THE data container in Java NIO APIs. Those who need to use features provided only by the NIO APIs (like non-blocking sockets) are doomed to use `ByteBuffer` for data transferring. Those who need to achieve better performance or use IO interfaces unavailable in Java StdLib will end up using libraries that might roll out their own data containers but usually still allowing to wrap or directly use `ByteBuffer` (like Netty or Aeron does).

It's possible to wrap a heap-allocated byte array (the backing storage for `kotlinx-io` segments) into a `HeapByteBuffer`, but the use of heap buffers comes with a cost. The majority of NIO API calls eventually perform a native call. If such a call (for example, a native wrapper for POSIX `write`) needs data, then NIO will supply it in the form of DirectByteBuffer or a memory address extracted from the DirectByteBuffer. If a user had provided `DirectByteBuffer`, then that buffer will be used, but if it was a `HeapByteBuffer`, then its content will be copied into an internal cached `DirectByteBuffer` instance and only then passed to the native API. If the buffer is empty, then the copying cost could be neglected, but as the buffer grows, it starts playing a more significant role in overall performance.

Besides performance issues with NIO API, a buffer residing in native memory is a necessity when it comes to implementing Java API for not yet supported native IO APIs such as io_uring, send w/ MSG_ZEROCOPY flag, epoll in the edge-triggering mode, etc. The only available option for allocating such a buffer and using it in a wide range of JVM versions supported by the Kotlin is by using `DirectByteBuffer`.

Unfortunately, using direct byte buffers is not always an option:
- some APIs don't directly support it on JVM (like `MessageDigest`)
- manipulations with the buffer itself works significantly slower on Android

So the only viable option might be to support both byte-arrays and `ByteBuffer`s as a backing storage and provide a way to choose what particular implementation to use when starting an app.

Tasks:
- [x] investigate ByteBuffers advantages/need to support it in `kotlinx-io`
- [ ] publish results of BB performance investigation
- [x] evaluate `kotlinx-io` performance with `DirectByteBuffer`
- [ ] publish performance characteristics of `kotlinx-io` w/ BB as a backing storage on JVM
- [ ] refactor the library to allow using different `Segment` implementations
- [ ] implement `DirectByteBuffer`-backed segments
- [ ] investigate JDK22 `MemorySegment`s usage instead of BB
- [ ] implement polymorphic segment
- [ ] port some benchmarks to Android
- [ ] evaluate baseline performance on Android
- [ ] evaluate DirectByteBuffers performance on Android
- [ ] investigate R8 features/capabilities/issues
- [ ] finalize and publish a design
- [ ] test-library support for multiple segment types
- [ ] tune performance (rewrite UTF8-manipulation routines, for example)

**Comment by fzhinkin** [comment]
The https://github.com/Kotlin/kotlinx-io/issues/135 will be done in the context of this project (at least partially).

**Comment by VDostoyevskiy** [comment]
“manipulations with the buffer itself works significantly slower on Android“
DirectByteBuffer is actually non-movable bytearray allocated on dalvik heap on Android platform,still it will never be copied when GC so we don't need to pay an extra copy for native IO. Any benchmark indicates that we should not use DirectByteBuffer on Android?
@fzhinkin

**Comment by fzhinkin** [comment]
Slowly processing the task list.

> publish results of BB performance investigation

Benchmarking results published here: https://github.com/fzhinkin/kotlinx-io-supplementary-benchmarks#kotlinx-io-supplementary-benchmarks

tl;dr
On JVM, writing or reading direct byte buffer via channel is usually faster then corresponding operation involving byte arrays and java.io streams.
The unexpected twist: on Android, byte buffer-based operations are ridiculously fast compared to their array-based java-io counterparts.

**Comment by fzhinkin** [comment]
I've also checked if having multiple segment types will affect the performance if only the one type is actually in use (the assumption is that at least the JVM will employ the CHA to avoid redundant type checks).

There's a branch (that won't compile to any target except JVM) [private/polymorphic-segments](https://github.com/Kotlin/kotlinx-io/tree/private/polymorphic-segments) where Segment was turned into an abstract class with two implementations - one with ByteArray inside (based on the `private/segments-public-api` branch) and another with the ByteBuffer inside (based on `private/dbb-benchmarking` branch). For the benchmarking purposes, ByteBuffer-backed segments were never loaded during the experiments (verified with class loading logs).

I won't post a large table as above, will just briefly summarize results:
- on JVM, there is no significant difference between results gathered for `private/segments-public-api` and `private/polymorphic-segments` branches: that's good, presence of ByteBuffer-backed segments won't affect those who don't need them;
- on Android, the situation is different: use of polymorphic segments makes performance worse. It's true even w/ R8 applied with a config that allows treating ByteBuffer-backed segments allocation path unreachable.

JVM benchmarking results are [here](https://github.com/Kotlin/kotlinx-io/tree/design/dbb/docs/design/byte-buffers/benchmarking/core/jvm-2024-01-26) and Android benchmarking results are [here](https://github.com/Kotlin/kotlinx-io/tree/design/dbb/docs/design/byte-buffers/benchmarking/core/android-2024-02-01).

**Comment by fzhinkin** [comment]
With all that being said about the performance aspect of ByteBuffers support, it's also worth mentioning that ByteBuffers on JVM and native-pointer-based segments on native would help with supporting memory-mapped files.
With array-backed segments only, memory-mapped files would require an additional class/interface.
With polymorphic segments, we could (not without caveats) wrap a ByteBuffer or mmaped ptr into a segment as a whole.

**Comment by fzhinkin** [comment]
Probably we can support ByteBuffers on JVM without hurting performance on Android by publishing a multi-release jar with a baseline implementation remaining the same (byte-array backed) but with polymorphic segments and BB-support enabled for, let's say, JDK9 and onwards. Android tooling ignores MRJ-stuff while dexing, so the trick might work. 👿 

I don't think it's a solution we should/could stick to, but that could solve an issue.

**Comment by rnett** [comment]
I see you have
> investigate JDK22 MemorySegments usage instead of BB

in the steps, did that ever go anywhere?  Even if using it as a backing storage by default doesn't work out, some kind of integration would be helpful to allow using kotlinx.io when working with native libraries on the JVM (using the project panama methods that use MemorySegment).

**Comment by Chuckame** [comment]
Hello there, bumping up this issue, in avro4k, have have this issue https://github.com/avro-kotlin/avro4k/issues/476 which legitimately ask for zero-copy deserialization. I would love that kotlinx-io provides a way to wrap a ByteBuffer as a Source. 

Or did you already consider to open the Source interface to allow user-specific code to make custom implementations (backed by BB, memory segment, unsafe, or any other funny stuff)? This would allow so much flexibility especially for decoding part.
